README.md
# XFYun NLS

[![js-standard-style](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://github.com/feross/standard)
[![action][ci-image]][ci-url]
[![license][license-image]][repository-url]
[![NPM version][npm-image]][npm-url]
[![npm download][download-image]][download-url]

[ci-image]: https://img.shields.io/github/workflow/status/funnyzak/xfyun-nls/Node.js%20CI
[ci-url]: https://github.com/funnyzak/xfyun-nls/actions
[license-image]: https://img.shields.io/github/license/funnyzak/xfyun-nls.svg?style=flat-square
[repository-url]: https://github.com/funnyzak/xfyun-nls
[npm-image]: https://img.shields.io/npm/v/xfyun-nls.svg?style=flat-square
[npm-url]: https://npmjs.org/package/xfyun-nls
[download-image]: https://img.shields.io/npm/dm/xfyun-nls.svg?style=flat-square
[download-url]: https://npmjs.org/package/xfyun-nls

讯飞云智能语音处理 Node 模块。

## 目录

- [XFYun NLS](#xfyun-nls)
  - [目录](#目录)
  - [开始](#开始)
  - [在线语音合成](#在线语音合成)
    - [用例](#用例)
    - [函数](#函数)
      - [`checkConfig(): Promise<boolean>`](#checkconfig-promiseboolean)
      - [`send(text: string, options?: XFYunTTS.BusinessOption): Promise<XFYunTTS.TTSComplete>`](#sendtext-string-options-xfyunttsbusinessoption-promisexfyunttsttscomplete)
    - [定义](#定义)
      - [XFYunTTS.BusinessOption](#xfyunttsbusinessoption)
      - [XFYunTTS.TTSComplete](#xfyunttsttscomplete)
  - [Author](#author)
  - [参考](#参考)
  - [License](#license)

## 开始

from [npm](https://github.com/npm/npm)

    $ npm install xfyun-nls

## 在线语音合成

### 用例

```js
const { XFYunTTS } = require('xfyun-nls');
const path = require('path');
const fs = require('fs');

!(async () => {
  const cachePath = path.join(process.cwd(), new Date().getTime().toString());
  fs.mkdirSync(cachePath);

  const _xfyTTS = new XFYunTTS(
    {
      appId: '5e3b7ffe',
      apiSecret: '8440c6f3dc6e3b864a043dbb6c37afd0',
      apiKey: '1394021f9ef53a9a9400b4d8a5f4652a',
      host: 'tts-api.xfyun.cn',
      uri: '/v2/tts',
      hostUrl: 'wss://tts-api.xfyun.cn/v2/tts'
    } /** XFYunTTS.Config **/,
    cachePath /** catchPath (optional) **/
  );

  // send
  console.log('TTS COMPLETE:', JSON.stringify(await _xfyTTS.send('你好')));
  // more ...
})();
```

了解更多 [TTS Define](https://github.com/funnyzak/xfyun-nls/blob/master/nls.d.ts).

### 函数

`XFYunTTS` 有如下方法:

#### `checkConfig(): Promise<boolean>`

返回值 `Promise<boolean>` - 返回讯飞云语音配置密钥是否有效。

```js
const checkRlt = await _XFYunTTS.checkConfig();
console.log(checkRlt ? 'the config is passed' : 'error config');
```

#### `send(text: string, options?: XFYunTTS.BusinessOption): Promise<XFYunTTS.TTSComplete>`

- `text` string - 要转换的文本。
- `options` XFYunTTS.BusinessOption (optional) - 合成设置。

返回值 `Promise<XFYunTTS.TTSComplete>` - 转换完成对象。

### 定义

#### XFYunTTS.BusinessOption

合成语言的业务参数，有如下参数：

- `tte` string - 文本编码格式。
- `aux` string - 音频编码。
- `vcn` string - 发音人。

更多请看定义 [BusinessOption](https://github.com/funnyzak/xfyun-nls/blob/master/nls.d.ts)。

[官方文档参考](https://www.xfyun.cn/doc/tts/online_tts/API.html#%E6%8E%A5%E5%8F%A3%E8%A6%81%E6%B1%820)。

#### XFYunTTS.TTSComplete

合成结果对象，有如下属性：

- `filePath` string - 音频文件最终保存路径。
- `suffix` string - 文件后缀。
- `rootPath` string - 所在父文件夹路径。
- `text` string - 合成的原文本。
- `options` XFYunTTS.BusinessOption - 合成的高级选项。
- `startTime` number - 合成任务开始的时间戳（毫秒）。
- `elapsed` number - 合成所耗费的时间（毫秒）。

更多请看定义 [TTSComplete](https://github.com/funnyzak/xfyun-nls/blob/master/nls.d.ts)。

## Author

| [![twitter/funnyzak](https://s.gravatar.com/avatar/c2437e240644b1317a4a356c6d6253ee?s=70)](https://twitter.com/funnyzak 'Follow @funnyzak on Twitter') |
| ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| [funnyzak](https://yycc.me/)                                                                                                                           |

## 参考

- [文本语音合成文档](https://www.xfyun.cn/doc/tts/online_tts/API.html)
- [错误慢参考解决](https://www.xfyun.cn/document/error-code)

## License

Apache-2.0 License © 2021 [funnyzak](https://github.com/funnyzak)


CHANGELOG.md
## 1.0.0 (2021-11-21)

Feature:

- 正式发布


test/nls.test.js
'use strict';

const expect = require('expect.js');
const XFYunTTS = require('../lib/nls');

describe('XFYunTTS Core', function () {
  describe('XFYunTTS', function () {
    it('should pass into "config"', function () {
      expect(function () {
        new XFYunTTS();
      }).to.throwException(/must pass "config"/);
    });
  });
});


test/tts.test.js
'use strict';

const expect = require('expect.js');
const XFYunTTS = require('../lib/tts');

describe('XFYunTTS Core', function () {
  describe('XFYunTTS', function () {
    it('should pass into "config"', function () {
      expect(function () {
        new XFYunTTS();
      }).to.throwException(/must pass "config"/);
    });
  });
});


lib/tts.ts
import WebSocket from 'ws';
// Type definitions for [~THE LIBRARY NAME~] [~OPTIONAL VERSION NUMBER~]
// Project: [~THE PROJECT NAME~]
// Definitions by: [~YOUR NAME~] <[~A URL FOR YOU~]>

/*~ This is the module template file for class modules.
 *~ You should rename it to index.d.ts and place it in a folder with the same name as the module.
 *~ For example, if you were writing a file for "super-greeter", this
 *~ file should be 'super-greeter/index.d.ts'
 */

/*~ Note that ES6 modules cannot directly export class objects.
 *~ This file should be imported using the CommonJS-style:
 *~   import x = require('someLibrary');
 *~
 *~ Refer to the documentation to understand common
 *~ workarounds for this limitation of ES6 modules.
 */

/*~ This declaration specifies that the class constructor function
 *~ is the exported object from the file
 */

declare class XFYunTTS {
  /**
   * 讯飞端配置
   *
   * @type {XFYunTTS.Config}
   * @memberof XFYunTTS
   */
  config: XFYunTTS.Config;

  /**
   * 合成语音缓存路径
   * @type {string}
   * @memberof XFYunTTS
   */
  cachePath: string;

  /**
   * 默认转换配置
   *
   * @type {XFYunTTS.BusinessOption}
   * @memberof XFYunTTS
   */
  defaultBusinessOption: XFYunTTS.BusinessOption;

  /**
   * 当前正在处理的转换配置
   *
   * @type {XFYunTTS.BusinessOption}
   * @memberof XFYunTTS
   */
  businessOption: XFYunTTS.BusinessOption;

  /**
   *长链接WS对象
   *
   * @type {WebSocket}
   * @memberof XFYunTTS
   */
  ws: WebSocket;

  /**
   * 构造函数
   * @param config 配置
   * @param cachePath 缓存路径
   */
  constructor(config: XFYunTTS.Config, cachePath?: string);

  /**
   * 打印
   * @param message
   * @param level info/warn/error
   */
  log(message: Object, level: string): void;

  /**
   * 获取当前WSS URL
   * @returns
   */
  getWssUrl(): string;

  /**
   * 合成结果
   * @param text 转换文本
   * @param options 合成选项
   */
  send(text, options: XFYunTTS.BusinessOption): Promise<XFYunTTS.TTSComplete>;

  /**
   *检查配置
   * @returns
   */
  checkConfig(): Promise<boolean>;

  /**
   * 鉴权签名
   * @param date RTC时间
   */
  signature(date: string): string;
}

declare namespace XFYunTTS {
  export interface Config {
    appId: string;
    apiSecret: string;
    apiKey: string;
    host: string;
    uri: string;
    hostUrl?: string;
  }

  /**
   * 转换结果
   */
  export interface TTSComplete {
    /**
     * 保存路径
     */
    filePath: string;

    /**
     * 文件后缀
     */
    suffix: string;

    /**
     * 所在父文件夹
     */
    rootPath: string;

    /**
     * 文本
     */
    text: string;

    /**
     * 转换选项
     */
    options: BusinessOption;

    /**
     * 开始时间（毫秒时间戳）
     */
    startTime: number;

    /**
     * 耗时（毫秒）
     */
    elapsed: number;
  }

  /**
   * 合成设置
   * https://www.xfyun.cn/doc/tts/online_tts/API.html#%E6%8E%A5%E5%8F%A3%E8%A6%81%E6%B1%82
   */
  export interface BusinessOption {
    /**
     * 文本编码格式
     * GB2312
     * GBK
     * BIG5
     * UNICODE(小语种必须使用UNICODE编码，合成的文本需使用utf16小端的编码方式，详见java示例demo)
     * GB18030
     * UTF8（小语种）
     */
    tte: 'UTF8' | 'GB2312' | 'GBK' | 'BIG5' | 'UNICODE' | 'GB18030';
    /**
     * 音频编码，可选值：
     * raw：未压缩的pcm
     * lame：mp3 (当aue=lame时需传参sfl=1)
     * speex-org-wb;7： 标准开源speex（for speex_wideband，即16k）数字代表指定压缩等级（默认等级为8）
     * speex-org-nb;7： 标准开源speex（for speex_narrowband，即8k）数字代表指定压缩等级（默认等级为8）
     * speex;7：压缩格式，压缩等级1~10，默认为7（8k讯飞定制speex）
     * speex-wb;7：压缩格式，压缩等级1~10，默认为7（16k讯飞定制speex）
     */
    aue: string;
    /**
     * 需要配合aue=lame使用，开启流式返回
     * mp3格式音频
     * 取值：1 开启
     */
    sfl?: number;
    /**
     * 音频采样率，可选值：
     * audio/L16;rate=8000：合成8K 的音频
     * audio/L16;rate=16000：合成16K 的音频
     * auf不传值：合成16K 的音频
     */
    auf?: string;
    /**
     *发音人，可选值：请到控制台添加试用或购买发音人，添加后即显示发音人参数值
     */
    vcn: string;
    /**
     * 语速，可选值：[0-100]，默认为50
     */
    speed?: number;
    /**
     * 音量，可选值：[0-100]，默认为50
     */
    volume?: number;
    /**
     * 音高，可选值：[0-100]，默认为50
     */
    pitch?: number;
    /**
     * 合成音频的背景音
     * 0:无背景音（默认值）
     * 1:有背景音
     */
    bgs?: number;
    /**
     * 设置英文发音方式：
     * 0：自动判断处理，如果不确定将按照英文词语拼写处理（缺省）
     * 1：所有英文按字母发音
     * 2：自动判断处理，如果不确定将按照字母朗读
     * 默认按英文单词发音
     */
    reg?: number;
    /**
     * 合成音频数字发音方式
     * 0：自动判断（默认值）
     * 1：完全数值
     * 2：完全字符串
     * 3：字符串优先
     */
    rdn?: string;
  }
}

export = XFYunTTS;


src/nls.js
'use strict';

const CryptoJS = require('crypto-js');
const WebSocket = require('ws');
const path = require('path');
const fs = require('fs');
const chalk = require('chalk');
const assert = require('assert');

class XFYunTTS {
  constructor(config, cachePath) {
    assert(config, 'must pass "config"');
    assert(config.appId, 'must pass "config.appId"');
    assert(config.apiSecret, 'must pass "config.apiSecret"');
    assert(config.apiKey, 'must pass "config.apiKey"');
    assert(config.host, 'must pass "config.host"');
    assert(config.uri, 'must pass "config.uri"');
    assert(config.hostUrl, 'must pass "config.hostUrl"');
    // assert(config.cachePath, 'must pass "cachePath"');
    // if (!fs.existsSync(cachePath)) {
    //   throw new Error(`"cachePath" - ${cachePath} is not exist.`);
    // }

    if (!cachePath) {
      cachePath = path.join(process.cwd(), new Date().getTime());
      fs.mkdirSync(cachePath);
    }

    this.ws = null;
    this.config = config;
    this.cachePath = cachePath;
    this.defaultBusinessOption = {
      aue: 'lame',
      sfl: 1,
      auf: 'audio/L16;rate=16000',
      vcn: 'xiaoyan',
      speed: 50,
      volume: 50,
      pitch: 50,
      bgs: 0,
      tte: 'UTF8',
      rdn: '0'
    };

    /**
     * 当前处理对象的高级配置
     */
    this.businessOption = { ...this.defaultBusinessOption };

    /**
     * WS 数据接受次数
     */
    this.receiveCount = 0;
  }

  /**
   * 打印
   * @param {*} msg
   * @param {*} level
   */
  log(msg, level = 'info') {
    const message = typeof msg !== 'string' ? JSON.stringify(msg) : msg;

    const warning = (_message) => chalk`{yellow WARNING:} ${_message}`;
    const info = (_message) => chalk`{magenta INFO:} ${_message}`;
    const error = (_message) => chalk`{red ERROR:} ${_message}`;
    switch (level) {
      case 'error':
        console.info(error(message));
        break;
      case 'warn':
        console.warn(warning(message));
        break;
      default:
        console.info(info(message));
        break;
    }
  }

  /**
   * 当前时间 RFC1123格式
   * @returns
   */
  rtcNow() {
    return new Date().toUTCString();
  }

  /**
   * 获取当前WSS URL
   * @returns
   */
  getWssUrl() {
    const _rtcNow = this.rtcNow();
    const wssUrl = `${this.config.hostUrl}?authorization=${this.signature(
      _rtcNow
    )}&date=${_rtcNow}&host=${this.config.host}`;
    this.log(`Wss URL: ${wssUrl}.`);
    return wssUrl;
  }

  /**
 * 网络检查配置是否正确
 * @returns
 */
  async checkConfig() {
    try {
      await this.send('Hello World');
      return true;
    } catch {
      return false;
    }
  }

  /**
   * 接受音频流
   * @returns
   */
  receive(fileSavePath) {
    this.log(`will save audio path: ${fileSavePath}.`);

    return new Promise((resolve, reject) => {
      if (this.ws === null) {
        reject(new Error('ws is null.'));
        return;
      }

      this.ws.onmessage = (ev) => {
        this.log(`ws received data: ${ev.data}.`);

        const res = JSON.parse(ev.data);
        if (res.code !== 0) {
          reject(new Error(`${res.code}: ${res.message}.`));
          this.ws.close();
          return;
        }

        const { audio } = res.data;
        if (audio !== null) {
          const audioBuf = Buffer.from(audio, 'base64');
          fs.writeFile(fileSavePath, audioBuf, { flag: 'a' }, (err) => {
            if (err) {
              reject(err);
              return;
            }
            this.log(`The file has been append!`);
          });
          this.receiveCount++;
        }

        if (res.code === 0 && res.data.status === 2) {
          this.ws.close();
          if (this.receiveCount > 0) {
            this.log(`file has been saved to path: ${fileSavePath}`);
            resolve(fileSavePath);
          } else {
            reject(new Error('receive error, maybe check speaker type.'));
          }
        }
      };
    });
  }

  /**
   * 重置配置
   */
  reset() {
    this.ws = new WebSocket(this.getWssUrl());
    this.receiveCount = 0;
    this.businessOption = { ...this.defaultBusinessOption };
  }

  /**
   * 合成语音
   * @param {*} text 合成文本
   * @param {*} options 高级业务设置
   * @returns
   */
  send(text, options = {}) {
    this.reset();

    this.businessOption = {
      ...this.defaultBusinessOption,
      ...options
    };

    return new Promise((resolve, reject) => {
      const _startTS = new Date().getTime();

      this.ws.onopen = () => {
        this.log(`task => text: ${text}, options: ${JSON.stringify(this.businessOption)}`);

        const frame = {
          common: {
            app_id: this.config.appId
          },
          business: { ...this.businessOption },
          data: {
            text: Buffer.from(text).toString('base64'),
            status: 2
          }
        };
        this.log(`ws send: ${JSON.stringify(frame)}`);
        this.ws.send(JSON.stringify(frame));
      };

      this.ws.onclose = () => {
        this.log('ws closed.');
      };

      this.ws.onerror = (ev) => {
        this.log(`ws error. data: ${JSON.stringify(ev)}`);
        reject(new Error(`ws onerror handle. data: ${JSON.stringify(ev)}`));
      };

      const filePath = this.randomSavePath();
      this.receive(filePath).then(() => {
        resolve({
          filePath,
          text,
          startTime: _startTS,
          options: this.businessOption,
          rootPath: this.cachePath,
          suffix: path.extname(filePath).substring(1),
          elapsed: new Date().getTime() - _startTS
        })
      }).catch(err => {
        this.log(`receive error: ${err.message}.`, 'error');
        reject(err);
      });
    });
  }

  /**
   * 随机保存路径
   * @returns
   */
  randomSavePath() {
    return path.join(
      this.cachePath,
      `${new Date().getTime().toString()}.${this.businessOption.aue === 'raw'
        ? 'pcm'
        : this.businessOption.aue === 'lame'
          ? 'mp3'
          : this.businessOption.aux.startsWith('speex')
            ? 'speex'
            : 'audio'
      }`
    );
  }

  /**
   * @param date  鉴权签名
   * @returns
   */
  signature(date) {
    const signatureOrigin = `host: ${this.config.host}\ndate: ${date}\nGET ${this.config.uri} HTTP/1.1`;
    const signatureSha = CryptoJS.HmacSHA256(
      signatureOrigin,
      this.config.apiSecret
    );
    const signature = CryptoJS.enc.Base64.stringify(signatureSha);
    const authorizationOrigin = `api_key="${this.config.apiKey}", algorithm="hmac-sha256", headers="host date request-line", signature="${signature}"`;
    const authStr = CryptoJS.enc.Base64.stringify(
      CryptoJS.enc.Utf8.parse(authorizationOrigin)
    );
    return authStr;
  }
}

module.exports = XFYunTTS;


.eslintrc.js
module.exports = {
  root: true,
  rules: {
    // 代码长度
    'max-len': [
      'error',
      {
        code: 1000,
        ignoreComments: true,
        ignoreUrls: true,
        ignoreTemplateLiterals: true
      }
    ],
    'semi': ['error', 'always'],
    'comma-dangle': ['error', 'never'],
    'import/no-unresolved': 'off',
    'import/no-extraneous-dependencies': 'off',
    'no-console': process.env.NODE_ENV === 'production' ? 'warn' : 'off',
    'no-debugger': process.env.NODE_ENV === 'production' ? 'warn' : 'off',
    'no-plusplus': 'off',
    'camelcase': 'off',
    'import/extensions': 'off',
    'no-use-before-define': 'off',
    'no-var-requires': 'off',
    'import/no-commonjs': 'off',
    'no-continue': 'off',
    'no-param-reassign': 'off',
    // 控制逗号前后的空格
    'comma-spacing': [2, { before: false, after: true }],
    // 控制逗号在行尾出现还是在行首出现
    // http://eslint.org/docs/rules/comma-style
    'comma-style': [2, 'last'],
    // 圈复杂度
    'complexity': [2, 15],
    // 文件末尾强制换行
    'eol-last': 2,
    'no-redeclare': [2, { builtinGlobals: true }],
    // 方法表达式是否需要命名
    'func-names': 0,
    'no-alert': 0, //禁止使用alert confirm prompt
    'no-array-constructor': 2, //禁止使用数组构造器
    'no-bitwise': 0, //禁止使用按位运算符
    'no-caller': 1, //禁止使用arguments.caller或arguments.callee
    'no-catch-shadow': 2, //禁止catch子句参数与外部作用域变量同名
    'no-class-assign': 2, //禁止给类赋值
    'no-cond-assign': 2, //禁止在条件表达式中使用赋值语句
    'no-const-assign': 2, //禁止修改const声明的变量
    'no-constant-condition': 2, //禁止在条件中使用常量表达式 if(true) if(1)
    'no-dupe-keys': 2, //在创建对象字面量时不允许键重复 {a:1,a:1}
    'no-dupe-args': 2, //函数参数不能重复
    'no-duplicate-case': 2, //switch中的case标签不能重复
    'no-else-return': 2, //如果if语句里面有return,后面不能跟else语句
    'quote-props': ['warn', 'consistent'],
    // 定义未使用
    'no-unused-vars': 'off',
    // 禁止使用 for..in  for..of
    'no-restricted-syntax': 1,
    // 禁止标识符下划线
    'no-underscore-dangle': 'off',
    // 导入多次
    'import/no-duplicates': 0,
    'import/prefer-default-export': 0,
    // 禁止promise 使用async
    'no-async-promise-executor': 0,
    // image元素必须有alt标签
    'jsx-a11y/alt-text': 0,
    // 变量作用域
    'no-shadow': 'off',
    // 不允许嵌套的三元表达式
    'no-nested-ternary': 'off',
    'no-await-in-loop': 'off',
    'class-methods-use-this': 'off',
    'prefer-promise-reject-errors': ['error', { allowEmptyReject: true }],
    // 对象/数组 解构
    'prefer-destructuring': [
      'error',
      {
        array: true,
        object: true
      },
      {
        enforceForRenamedProperties: false
      }
    ]
  },
  env: {
    es6: true,
    node: true,
    browser: true
  },
  globals: {
    describe: true,
    it: true,
    before: true,
    after: true
  },
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true
    }
  },
  extends: 'eslint:recommended'
};


lib/tts.d.ts
import WebSocket from 'ws';
// Type definitions for [~THE LIBRARY NAME~] [~OPTIONAL VERSION NUMBER~]
// Project: [~THE PROJECT NAME~]
// Definitions by: [~YOUR NAME~] <[~A URL FOR YOU~]>

/*~ This is the module template file for class modules.
 *~ You should rename it to index.d.ts and place it in a folder with the same name as the module.
 *~ For example, if you were writing a file for "super-greeter", this
 *~ file should be 'super-greeter/index.d.ts'
 */

/*~ Note that ES6 modules cannot directly export class objects.
 *~ This file should be imported using the CommonJS-style:
 *~   import x = require('someLibrary');
 *~
 *~ Refer to the documentation to understand common
 *~ workarounds for this limitation of ES6 modules.
 */

/*~ This declaration specifies that the class constructor function
 *~ is the exported object from the file
 */

declare class XFYunTTS {
  /**
   * 讯飞端配置
   *
   * @type {XFYunTTS.Config}
   * @memberof XFYunTTS
   */
  config: XFYunTTS.Config;

  /**
   * 合成语音缓存路径
   * @type {string}
   * @memberof XFYunTTS
   */
  cachePath: string;

  /**
   * 默认转换配置
   *
   * @type {XFYunTTS.BusinessOption}
   * @memberof XFYunTTS
   */
  defaultBusinessOption: XFYunTTS.BusinessOption;

  /**
   * 当前正在处理的转换配置
   *
   * @type {XFYunTTS.BusinessOption}
   * @memberof XFYunTTS
   */
  businessOption: XFYunTTS.BusinessOption;

  /**
   *长链接WS对象
   *
   * @type {WebSocket}
   * @memberof XFYunTTS
   */
  ws: WebSocket;

  /**
   * 构造函数
   * @param config 配置
   * @param cachePath 缓存路径
   */
  constructor(config: XFYunTTS.Config, cachePath?: string);

  /**
   * 打印
   * @param message
   * @param level info/warn/error
   */
  log(message: Object, level: string): void;

  /**
   * 获取当前WSS URL
   * @returns
   */
  getWssUrl(): string;

  /**
   * 合成结果
   * @param text 转换文本
   * @param options 合成选项
   */
  send(text, options: XFYunTTS.BusinessOption): Promise<XFYunTTS.TTSComplete>;

  /**
   *检查配置
   * @returns
   */
  checkConfig(): Promise<boolean>;

  /**
   * 鉴权签名
   * @param date RTC时间
   */
  signature(date: string): string;
}

declare namespace XFYunTTS {
  export interface Config {
    appId: string;
    apiSecret: string;
    apiKey: string;
    host: string;
    uri: string;
    hostUrl?: string;
  }

  /**
   * 转换结果
   */
  export interface TTSComplete {
    /**
     * 音频文件最终保存路径
     */
    filePath: string;

    /**
     * 文件后缀
     */
    suffix: string;

    /**
     * 所在父文件夹
     */
    rootPath: string;

    /**
     * 文本
     */
    text: string;

    /**
     * 转换选项
     */
    options: BusinessOption;

    /**
     * 开始时间（毫秒时间戳）
     */
    startTime: number;

    /**
     * 耗时（毫秒）
     */
    elapsed: number;
  }

  /**
   * 合成设置
   * https://www.xfyun.cn/doc/tts/online_tts/API.html#%E6%8E%A5%E5%8F%A3%E8%A6%81%E6%B1%82
   */
  export interface BusinessOption {
    /**
     * 文本编码格式
     * GB2312
     * GBK
     * BIG5
     * UNICODE(小语种必须使用UNICODE编码，合成的文本需使用utf16小端的编码方式，详见java示例demo)
     * GB18030
     * UTF8（小语种）
     */
    tte: 'UTF8' | 'GB2312' | 'GBK' | 'BIG5' | 'UNICODE' | 'GB18030';
    /**
     * 音频编码，可选值：
     * raw：未压缩的pcm
     * lame：mp3 (当aue=lame时需传参sfl=1)
     * speex-org-wb;7： 标准开源speex（for speex_wideband，即16k）数字代表指定压缩等级（默认等级为8）
     * speex-org-nb;7： 标准开源speex（for speex_narrowband，即8k）数字代表指定压缩等级（默认等级为8）
     * speex;7：压缩格式，压缩等级1~10，默认为7（8k讯飞定制speex）
     * speex-wb;7：压缩格式，压缩等级1~10，默认为7（16k讯飞定制speex）
     */
    aue: string;
    /**
     * 需要配合aue=lame使用，开启流式返回
     * mp3格式音频
     * 取值：1 开启
     */
    sfl?: number;
    /**
     * 音频采样率，可选值：
     * audio/L16;rate=8000：合成8K 的音频
     * audio/L16;rate=16000：合成16K 的音频
     * auf不传值：合成16K 的音频
     */
    auf?: string;
    /**
     *发音人，可选值：请到控制台添加试用或购买发音人，添加后即显示发音人参数值
     */
    vcn: string;
    /**
     * 语速，可选值：[0-100]，默认为50
     */
    speed?: number;
    /**
     * 音量，可选值：[0-100]，默认为50
     */
    volume?: number;
    /**
     * 音高，可选值：[0-100]，默认为50
     */
    pitch?: number;
    /**
     * 合成音频的背景音
     * 0:无背景音（默认值）
     * 1:有背景音
     */
    bgs?: number;
    /**
     * 设置英文发音方式：
     * 0：自动判断处理，如果不确定将按照英文词语拼写处理（缺省）
     * 1：所有英文按字母发音
     * 2：自动判断处理，如果不确定将按照字母朗读
     * 默认按英文单词发音
     */
    reg?: number;
    /**
     * 合成音频数字发音方式
     * 0：自动判断（默认值）
     * 1：完全数值
     * 2：完全字符串
     * 3：字符串优先
     */
    rdn?: string;
  }
}

export = XFYunTTS;


src/tts.js
'use strict';

const CryptoJS = require('crypto-js');
const WebSocket = require('ws');
const path = require('path');
const fs = require('fs');
const chalk = require('chalk');
const assert = require('assert');

class XFYunTTS {
  constructor(config, cachePath) {
    assert(config, 'must pass "config"');
    assert(config.appId, 'must pass "config.appId"');
    assert(config.apiSecret, 'must pass "config.apiSecret"');
    assert(config.apiKey, 'must pass "config.apiKey"');
    assert(config.host, 'must pass "config.host"');
    assert(config.uri, 'must pass "config.uri"');
    assert(config.hostUrl, 'must pass "config.hostUrl"');
    // assert(config.cachePath, 'must pass "cachePath"');
    // if (!fs.existsSync(cachePath)) {
    //   throw new Error(`"cachePath" - ${cachePath} is not exist.`);
    // }

    if (!cachePath) {
      cachePath = path.join(process.cwd(), new Date().getTime());
      fs.mkdirSync(cachePath);
    }

    this.ws = null;
    this.config = config;
    this.cachePath = cachePath;
    this.defaultBusinessOption = {
      aue: 'lame',
      sfl: 1,
      auf: 'audio/L16;rate=16000',
      vcn: 'xiaoyan',
      speed: 50,
      volume: 50,
      pitch: 50,
      bgs: 0,
      tte: 'UTF8',
      rdn: '0'
    };

    /**
     * 当前处理对象的高级配置
     */
    this.businessOption = { ...this.defaultBusinessOption };

    /**
     * WS 数据接受次数
     */
    this.receiveCount = 0;
  }

  /**
   * 打印
   * @param {*} msg
   * @param {*} level
   */
  log(msg, level = 'info') {
    const message = typeof msg !== 'string' ? JSON.stringify(msg) : msg;

    const warning = (_message) => chalk`{yellow WARNING:} ${_message}`;
    const info = (_message) => chalk`{magenta INFO:} ${_message}`;
    const error = (_message) => chalk`{red ERROR:} ${_message}`;
    switch (level) {
      case 'error':
        console.info(error(message));
        break;
      case 'warn':
        console.warn(warning(message));
        break;
      default:
        console.info(info(message));
        break;
    }
  }

  /**
   * 当前时间 RFC1123格式
   * @returns
   */
  rtcNow() {
    return new Date().toUTCString();
  }

  /**
   * 获取当前WSS URL
   * @returns
   */
  getWssUrl() {
    const _rtcNow = this.rtcNow();
    const wssUrl = `${this.config.hostUrl}?authorization=${this.signature(
      _rtcNow
    )}&date=${_rtcNow}&host=${this.config.host}`;
    this.log(`Wss URL: ${wssUrl}.`);
    return wssUrl;
  }

  /**
   * 网络检查配置是否正确
   * @returns
   */
  async checkConfig() {
    try {
      await this.send('Hello World');
      return true;
    } catch (error) {
      this.log(`check config error: ${error.message}`, 'error');
      return false;
    }
  }

  /**
   * 接受音频流
   * @returns
   */
  receive(fileSavePath) {
    this.log(`will save audio path: ${fileSavePath}.`);

    return new Promise((resolve, reject) => {
      if (this.ws === null) {
        reject(new Error('ws is null.'));
        return;
      }

      this.ws.onmessage = (ev) => {
        this.log(`ws received data: ${ev.data}.`);

        const res = JSON.parse(ev.data);
        if (res.code !== 0) {
          reject(new Error(`${res.code}: ${res.message}.`));
          this.ws.close();
          return;
        }

        const { audio } = res.data;
        if (audio !== null) {
          const audioBuf = Buffer.from(audio, 'base64');
          fs.writeFile(fileSavePath, audioBuf, { flag: 'a' }, (err) => {
            if (err) {
              reject(err);
              return;
            }
            this.log(`The file has been append!`);
          });
          this.receiveCount++;
        }

        if (res.code === 0 && res.data.status === 2) {
          this.ws.close();
          if (this.receiveCount > 0) {
            this.log(`file has been saved to path: ${fileSavePath}`);
            resolve(fileSavePath);
          } else {
            reject(new Error('receive error, maybe check speaker type.'));
          }
        }
      };
    });
  }

  /**
   * 重置配置
   */
  reset() {
    this.ws = new WebSocket(this.getWssUrl());
    this.receiveCount = 0;
    this.businessOption = { ...this.defaultBusinessOption };
  }

  /**
   * 合成语音
   * @param {*} text 合成文本
   * @param {*} options 高级业务设置
   * @returns
   */
  send(text, options = {}) {
    this.reset();

    this.businessOption = {
      ...this.defaultBusinessOption,
      ...options
    };

    return new Promise((resolve, reject) => {
      const _startTS = new Date().getTime();

      this.ws.onopen = () => {
        this.log(
          `task => text: ${text}, options: ${JSON.stringify(
            this.businessOption
          )}`
        );

        const frame = {
          common: {
            app_id: this.config.appId
          },
          business: { ...this.businessOption },
          data: {
            text: Buffer.from(text).toString('base64'),
            status: 2
          }
        };
        this.log(`ws send: ${JSON.stringify(frame)}`);
        this.ws.send(JSON.stringify(frame));
      };

      this.ws.onclose = () => {
        this.log('ws closed.');
      };

      this.ws.onerror = (ev) => {
        const errorMsg = `ws error. maybe 1.key setting error. 2. vcn permission`;
        this.log(errorMsg, 'error');
        reject(new Error(errorMsg));
      };

      const filePath = this.randomSavePath();
      this.receive(filePath)
        .then(() => {
          resolve({
            filePath,
            text,
            startTime: _startTS,
            options: this.businessOption,
            rootPath: this.cachePath,
            suffix: path.extname(filePath).substring(1),
            elapsed: new Date().getTime() - _startTS
          });
        })
        .catch((err) => {
          this.log(`receive error: ${err.message}.`, 'error');
          reject(err);
        });
    });
  }

  /**
   * 随机保存路径
   * @returns
   */
  randomSavePath() {
    return path.join(
      this.cachePath,
      `${new Date().getTime().toString()}.${
        this.businessOption.aue === 'raw'
          ? 'pcm'
          : this.businessOption.aue === 'lame'
          ? 'mp3'
          : this.businessOption.aux.startsWith('speex')
          ? 'speex'
          : 'audio'
      }`
    );
  }

  /**
   * @param date  鉴权签名
   * @returns
   */
  signature(date) {
    const signatureOrigin = `host: ${this.config.host}\ndate: ${date}\nGET ${this.config.uri} HTTP/1.1`;
    const signatureSha = CryptoJS.HmacSHA256(
      signatureOrigin,
      this.config.apiSecret
    );
    const signature = CryptoJS.enc.Base64.stringify(signatureSha);
    const authorizationOrigin = `api_key="${this.config.apiKey}", algorithm="hmac-sha256", headers="host date request-line", signature="${signature}"`;
    const authStr = CryptoJS.enc.Base64.stringify(
      CryptoJS.enc.Utf8.parse(authorizationOrigin)
    );
    return authStr;
  }
}

module.exports = XFYunTTS;
